mm2src/coins/eth/eth_tests.rs

```diff
@@ -1342,3 +1342,117 @@ fn test_sign_verify_message() {
         .unwrap();
     assert!(is_valid);
 }
+
+#[test]
+fn test_eth_extract_secret() {
+    let key_pair = KeyPair::from_secret_slice(
+        &hex::decode("809465b17d0a4ddb3e4c69e8f23c2cabad868f51f8bed5c765ad1d6516c3306f").unwrap(),
+    )
+    .unwrap();
+    let transport = Web3Transport::single_node("https://ropsten.infura.io/v3/c01c1b4cf66642528547624e1d6d9d6b", false);
+    let web3 = Web3::new(transport);
+    let ctx = MmCtxBuilder::new().into_mm_arc();
+
+    let swap_contract_address = Address::from("0x7Bc1bBDD6A0a722fC9bffC49c921B685ECB84b94");
+    let coin = EthCoin(Arc::new(EthCoinImpl {
+        coin_type: EthCoinType::Erc20 {
+            platform: "ETH".to_string(),
+            token_addr: Address::from("0xc0eb7aed740e1796992a08962c15661bdeb58003"),
+        },
+        decimals: 18,
+        gas_station_url: None,
+        gas_station_decimals: ETH_GAS_STATION_DECIMALS,
+        gas_station_policy: GasStationPricePolicy::MeanAverageFast,
+        history_sync_state: Mutex::new(HistorySyncState::NotEnabled),
+        my_address: key_pair.address(),
+        sign_message_prefix: Some(String::from("Ethereum Signed Message:\n")),
+        key_pair,
+        swap_contract_address,
+        fallback_swap_contract: None,
+        ticker: "ETH".into(),
+        web3_instances: vec![Web3Instance {
+            web3: web3.clone(),
+            is_parity: true,
+        }],
+        web3,
+        ctx: ctx.weak(),
+        required_confirmations: 1.into(),
+        chain_id: None,
+        logs_block_range: DEFAULT_LOGS_BLOCK_RANGE,
+        nonce_lock: new_nonce_lock(),
+        erc20_tokens_infos: Default::default(),
+    }));
+
+    // raw transaction bytes of https://ropsten.etherscan.io/tx/0xe18bbca69dea9a4624e1f5b0b2021d5fe4c8daa03f36084a8ba011b08e5cd938
```

Comment:
The mentioned transaction calls `erc20Payment` of swap contract, so `extract_secret` should not be applicable to it. Is there a mistake in the comment?

Reply:
Nop, the comment is not a mistake. I actually made use of the TX to generate the secret used for the test. I didn't know `extract_secret` isn't compatible with `swap contract`. Will make changes

Reply:
@artemii235 why do you think `extract_secret` isn't compatible with swap contract tx? 

implementation of `extract_secret` for eth..
```rust
    fn extract_secret(&self, _secret_hash: &[u8], spend_tx: &[u8]) -> Result<Vec<u8>, String> {
        let unverified: UnverifiedTransaction = try_s!(rlp::decode(spend_tx));
        let function = try_s!(SWAP_CONTRACT.function("receiverSpend"));
        let tokens = try_s!(function.decode_input(&unverified.data));
        if tokens.len() < 3 {
            return ERR!("Invalid arguments in 'receiverSpend' call: {:?}", tokens);
        }
        match &tokens[2] {
            Token::FixedBytes(secret) => Ok(secret.to_vec()),
            _ => ERR!(
                "Expected secret to be fixed bytes, decoded function data is {:?}",
                tokens
            ),
        }
    }
```

Reply:
The mentioned transaction https://ropsten.etherscan.io/tx/0xe18bbca69dea9a4624e1f5b0b2021d5fe4c8daa03f36084a8ba011b08e5cd938 calls `erc20Payment` smart contract function.
<img width="1040" alt="image" src="https://user-images.githubusercontent.com/15745003/194540775-1ca8a9a0-5c76-4ef1-b9c3-9335cd35634c.png">

While `extract_secret` should be called on the transaction calling `receiverSpend` (e.g. https://ropsten.etherscan.io/tx/0x432935be34424a0aa0403061652be8522da5b8637a6e499325ea910c430cd97d) - the name is actually hard-coded in `extract_secret`. It should have worked at all, but it somehow decodes the function input bytes and returns the *token address* padded with 12 zero bytes. There should be a bug in `decode_input` function - I added  it in my fork of `ethabi` crate. 

 cc @sergeyboyko0791

Reply:
Oh great, thanks 👍

Reply:
What's interesting: even the most recent official `ethabi` release doesn't seem to perform any validation that the input is actually from the original fn: https://github.com/rust-ethereum/ethabi/blob/6f18e11621d29b6cd978ef39b48fa572b823bb80/ethabi/src/function.rs#L78. The code expects input without 4 bytes short signature. So we have to implement additional validation in our code.